### ISSUE TITLE
Unexpected "Nullability of type argument doesn't match constraint." warnings when a type parameter with Oblivious constraint is used as a type argument

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#34892]
```
#nullable enable
public class AA
{
    public void M3<T3>(T3 z) where T3 : class { }

    public void M4<T4>(T4 z) where T4 : AA { }

#nullable disable
    public void F1<T1>(T1 x) where T1 : class
    {
#nullable enable
        M3<T1>(x); // warning CS8634: The type 'T1' cannot be used as type parameter 'T3' in the generic type or method 'AA.M3<T3>(T3)'. Nullability of type argument 'T1' doesn't match 'class' constraint.
    }

#nullable disable
    public void F2<T2>(T2 x) where T2 : AA
    {
#nullable enable
        M4<T2>(x); // warning CS8631: The type 'T2' cannot be used as type parameter 'T4' in the generic type or method 'AA.M4<T4>(T4)'. Nullability of type argument 'T2' doesn't match constraint type 'AA'.
    }
}
```

Observed:
```
warning CS8634: The type 'T1' cannot be used as type parameter 'T3' in the generic type or method 'AA.M3<T3>(T3)'. Nullability of type argument 'T1' doesn't match 'class' constraint.
warning CS8631: The type 'T2' cannot be used as type parameter 'T4' in the generic type or method 'AA.M4<T4>(T4)'. Nullability of type argument 'T2' doesn't match constraint type 'AA'.
```

Expected: No warnings.
